gusto/initialisation_tools.py

```diff
@@ -293,117 +305,170 @@ def moist_hydrostatic_balance(state, theta_e, water_t, pi_boundary=Constant(1.0)
-    p = thermodynamics.p(state.parameters, pi)
-    w_sat = thermodynamics.r_sat(state.parameters, T, p)
+    The general strategy is to split up the solving into two steps:
+    1) finding rho to balance the theta profile
+    2) finding theta_v and r_v to get back theta_d and H
+    We iteratively solve these steps until we (hopefully)
+    converge to a solution.
 
-    F = (-gamma * theta_e * dxp
-         + gamma * thermodynamics.theta_e(state.parameters, T, p, w_v, water_t) * dxp
-         - phi * w_v * dxp
-         + phi * w_sat * dxp
-         + cp * inner(v, w) * dxp
-         - cp * div(w * theta_v / (1.0 + water_t)) * pi * dxp
-         + psi * div(theta_v * v / (1.0 + water_t)) * dxp
-         + cp * inner(w, n) * pi_boundary * theta_v / (1.0 + water_t) * ds_b
-         + g * inner(w, state.k) * dxp)
+    :arg state: The :class:`State` object.
+    :arg theta_d: The initial dry potential temperature profile.
+    :arg H: The relative humidity profile.
+    :arg pi0: Optional function to put exner pressure into.
+    :arg top: If True, set a boundary condition at the top, otherwise
+              it will be at the bottom.
+    :arg pi_boundary: The value of pi on the specified boundary.
+    """
 
-    bcs = [DirichletBC(Z.sub(3), 0.0, "top")]
+    theta0 = state.fields('theta')
+    rho0 = state.fields('rho')
+    water_v0 = state.fields('water_v')
 
-    problem = NonlinearVariationalProblem(F, z, bcs=bcs)
-    solver = NonlinearVariationalSolver(problem, solver_parameters=params)
+    # Calculate hydrostatic Pi
+    Vt = theta0.function_space()
+    Vr = rho0.function_space()
+    R_d = state.parameters.R_d
+    R_v = state.parameters.R_v
+    epsilon = R_d / R_v
 
-    solver.solve()
+    VDG = state.spaces("DG")
+    if any(deg > 2 for deg in VDG.ufl_element().degree()):
+        state.logger.warning("default quadrature degree most likely not sufficient for this degree element")
 
-    theta_v, w_v, pi, v = z.split()
+    # apply first guesses
+    theta0.assign(theta_d * 1.01)
+    water_v0.assign(0.01)
+
+    rho_h = Function(Vr)
+    rho_averaged = Function(Vt)
+    rho_broken = Function(FunctionSpace(state.mesh, BrokenElement(Vt.ufl_element())))
+    rho_recoverer = Recoverer(rho_broken, rho_averaged)
+    w_h = Function(Vt)
+    delta = 1.0
+
+    # make expressions for determining water_v0
+    pie = thermodynamics.pi(state.parameters, rho_averaged, theta0)
+    p = thermodynamics.p(state.parameters, pie)
+    T = thermodynamics.T(state.parameters, theta0, pie, water_v0)
+    r_v_expr = thermodynamics.r_v(state.parameters, H, T, p)
+
+    # make expressions to evaluate residual
+    pi_ev = thermodynamics.pi(state.parameters, rho_averaged, theta0)
+    p_ev = thermodynamics.p(state.parameters, pi_ev)
+    T_ev = thermodynamics.T(state.parameters, theta0, pi_ev, water_v0)
+    RH_ev = thermodynamics.RH(state.parameters, water_v0, T_ev, p_ev)
+    RH = Function(Vt)
+
+    for i in range(40):
```

Comment:
My comments and worries about the first instance of this still applies. I recommend making sure all this fails gracefully when it does.

Reply:
Yes will do the same, thanks! (At least both methods are very similar...)